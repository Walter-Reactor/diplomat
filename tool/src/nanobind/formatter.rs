//! This module contains functions for formatting types

use crate::cpp::Cpp2Formatter;
use diplomat_core::hir::{self, DocsUrlGenerator, StringEncoding, TypeContext, TypeId};
use std::borrow::Cow;

/// This type mediates all formatting
///
/// All identifiers from the HIR should go through here before being formatted
/// into the output: This makes it easy to handle reserved words or add rename support
///
/// If you find yourself needing an identifier formatted in a context not yet available here, please add a new method
///
/// This type may be used by other backends attempting to figure out the names
/// of C types and methods.
pub(crate) struct PyFormatter<'tcx> {
    pub cxx: Cpp2Formatter<'tcx>,
}

impl<'tcx> PyFormatter<'tcx> {
    pub fn new(tcx: &'tcx TypeContext, docs_url_gen: &'tcx DocsUrlGenerator) -> Self {
        Self {
            cxx: Cpp2Formatter::new(tcx, docs_url_gen),
        }
    }

    /// Resolve and format the nested module names for this type
    /// Returns an iterator to the namespaces. Will always have at least one entry
    pub fn fmt_namespaces(&self, id: TypeId) -> impl Iterator<Item = &'tcx str> {
        let resolved = self.cxx.c.tcx().resolve_type(id);
        resolved
            .attrs()
            .namespace
            .as_ref()
            .map(|v| v.split("::"))
            .into_iter()
            .flatten()
    }

    /// Resolve the name of the module to use
    pub fn fmt_module(&'tcx self, id: TypeId, default: &'tcx str) -> Cow<'tcx, str> {
        self.fmt_namespaces(id).last().unwrap_or(default).into()
    }

    /// Resolve and format a named type for use in code (without the namespace)
    pub fn fmt_type_name_unnamespaced(&self, id: TypeId) -> Cow<'tcx, str> {
        self.cxx.fmt_type_name_unnamespaced(id)
    }

    /// Resolve and format a named type for use in code
    pub fn fmt_type_name(&self, id: TypeId) -> Cow<'tcx, str> {
        self.cxx.fmt_type_name(id)
    }

    /// Resolve and format the name of a type for use in header names
    pub fn fmt_decl_header_path(&self, id: TypeId) -> String {
        self.cxx.fmt_decl_header_path(id)
    }

    /// Resolve and format the name of a type for use in header names
    pub fn fmt_impl_file_path(&self, id: TypeId) -> String {
        self.cxx.fmt_impl_header_path(id)
    }

    /// Format an enum variant.
    pub fn fmt_enum_variant(&self, variant: &'tcx hir::EnumVariant) -> Cow<'tcx, str> {
        self.cxx.fmt_enum_variant(variant)
    }

    /// Format a field name or parameter name
    // might need splitting in the future if we decide to support renames here
    pub fn fmt_param_name<'a>(&self, ident: &'a str) -> Cow<'a, str> {
        self.cxx.fmt_param_name(ident)
    }

    pub fn fmt_c_type_name(&self, id: TypeId) -> Cow<'tcx, str> {
        self.cxx.fmt_c_type_name(id)
    }

    pub fn fmt_c_ptr<'a>(&self, ident: &'a str, mutability: hir::Mutability) -> Cow<'a, str> {
        self.cxx.fmt_c_ptr(ident, mutability)
    }

    pub fn fmt_borrowed<'a>(&self, ident: &'a str, mutability: hir::Mutability) -> Cow<'a, str> {
        self.cxx.fmt_borrowed(ident, mutability)
    }

    pub fn fmt_move_ref<'a>(&self, ident: &'a str) -> Cow<'a, str> {
        self.cxx.fmt_move_ref(ident)
    }

    pub fn fmt_optional_borrowed<'a>(
        &self,
        ident: &'a str,
        mutability: hir::Mutability,
    ) -> Cow<'a, str> {
        self.cxx.fmt_optional_borrowed(ident, mutability)
    }

    pub fn fmt_owned<'a>(&self, ident: &'a str) -> Cow<'a, str> {
        self.cxx.fmt_owned(ident)
    }

    pub fn fmt_borrowed_slice<'a>(
        &self,
        ident: &'a str,
        mutability: hir::Mutability,
    ) -> Cow<'a, str> {
        self.cxx.fmt_borrowed_slice(ident, mutability)
    }

    pub fn fmt_borrowed_str(&self, encoding: StringEncoding) -> Cow<'static, str> {
        self.cxx.fmt_borrowed_str(encoding)
    }

    /// Format a method
    pub fn fmt_method_name<'a>(&self, method: &'a hir::Method) -> Cow<'a, str> {
        self.cxx.fmt_method_name(method)
    }

    pub fn namespace_c_method_name(&self, ty: TypeId, name: &str) -> String {
        self.cxx.namespace_c_method_name(ty, name)
    }

    /// Get the primitive type as a C type
    pub fn fmt_primitive_as_c(&self, prim: hir::PrimitiveType) -> Cow<'static, str> {
        self.cxx.fmt_primitive_as_c(prim)
    }
}
