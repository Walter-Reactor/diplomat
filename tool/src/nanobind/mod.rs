mod binding;
mod formatter;
mod ty;

use std::{borrow::Cow, collections::HashSet, path::Path};

use crate::{ErrorStore, FileMap};
use binding::Binding;
use diplomat_core::hir::{self, BackendAttrSupport};
use formatter::PyFormatter;
use serde::{Deserialize, Serialize};
use ty::TyGenContext;

// Python support using the nanobind c++ library to create a python binding.
//
// The generated nanobind.cpp files requires linking with nanobind
// See the feature_test project for an example of a pyproject & CMakeLists.txt which can be compiled
// using pip install. Compilation requires a C++ compiler & CMake, as well as a downloaded
// copy of libnanobind.

pub(crate) fn attr_support() -> BackendAttrSupport {
    let mut a = BackendAttrSupport::default();

    a.namespacing = true;
    a.memory_sharing = true;
    a.non_exhaustive_structs = false;
    a.method_overloading = true;
    a.utf8_strings = true;
    a.utf16_strings = true;
    a.static_slices = true;

    a.constructors = false; // TODO
    a.named_constructors = false;
    a.fallible_constructors = false;
    a.accessors = false;
    a.comparators = false; // TODO
    a.stringifiers = false; // TODO
    a.iterators = false; // TODO
    a.iterables = false; // TODO
    a.indexing = false; // TODO
    a.option = true;
    a.callbacks = true;
    a.traits = false;

    a
}

#[derive(Debug, Clone, Deserialize, Serialize)]
struct PythonConfig {
    lib_name: String,
}

pub(crate) fn run<'tcx>(
    tcx: &'tcx hir::TypeContext,
    conf_path: Option<&Path>,
) -> (FileMap, ErrorStore<'tcx, String>) {
    let conf_path = conf_path.expect("Nanobind library needs to be called with config");

    let conf_str = std::fs::read_to_string(conf_path)
        .unwrap_or_else(|err| panic!("Failed to open config file {conf_path:?}: {err}"));
    let PythonConfig { lib_name } = toml::from_str::<PythonConfig>(&conf_str)
        .expect("Failed to parse config. Required field is 'lib_name'");

    let files = FileMap::default();
    let formatter = PyFormatter::new(tcx);
    let errors = ErrorStore::default();

    let nanobind_filepath = format!("{lib_name}_ext.cpp");
    let mut binding = Binding::new();
    binding.module_name = Cow::from(lib_name);

    let mut submodules = HashSet::<Cow<str>>::new();
    for (id, ty) in tcx.all_types() {
        if ty.attrs().disable {
            // Skip type if disabled
            continue;
        }

        let _type_name_unnamespaced = formatter.fmt_type_name(id);
        let decl_header_path = formatter.fmt_decl_header_path(id);
        let impl_file_path = formatter.fmt_impl_file_path(id);

        let mut context = TyGenContext {
            formatter: &formatter,
            errors: &errors,
            c2: crate::c::TyGenContext {
                tcx,
                formatter: &formatter.c,
                errors: &errors,
                is_for_cpp: false,
                id: id.into(),
                decl_header_path: &decl_header_path,
                impl_header_path: &impl_file_path,
            },
            binding: &mut binding,
            submodules: &mut submodules,
            generating_struct_fields: false,
        };

        context
            .binding
            .includes
            .insert(impl_file_path.clone().into());

        let guard = errors.set_context_ty(ty.name().as_str().into());
        match ty {
            hir::TypeDef::Enum(o) => context.gen_enum_def(o, id),
            hir::TypeDef::Opaque(o) => context.gen_opaque_def(o, id),
            hir::TypeDef::Struct(s) => context.gen_struct_def(s, id),
            hir::TypeDef::OutStruct(s) => context.gen_struct_def(s, id),
            _ => unreachable!("unknown AST/HIR variant"),
        }
        drop(guard);
    }
    files.add_file(nanobind_filepath.to_owned(), binding.to_string());

    // Write out wheel metadata files
    {
        // #[derive(Template)]
        // #[template(path = "nanobind/pyproject.toml.jinja", escape = "none")]
        // struct ImplTemplate<'a> {
        //     _ty: &'a hir::EnumDef,
        //     _fmt: &'a PyFormatter<'a>,
        //     type_name: &'a str,
        //     _ctype: &'a str,
        //     values: &'a [&'a EnumVariant],
        //     module: &'a str,
        //     modules: Vec<(Cow<'a, str>, Cow<'a, str>)>,
        // }

        // ImplTemplate {
        //     _ty: ty,
        //     _fmt: self.formatter,
        //     type_name: &type_name,
        //     _ctype: &ctype,
        //     values: values.as_slice(),
        //     module: self.formatter.fmt_module(id).borrow(),
        //     modules: self.get_module_defs(id, None),
        // }
        // .render_into(self.binding)
        // .unwrap();
    }
    (files, errors)
}

#[cfg(test)]
mod test {
    use diplomat_core::hir::{self, TypeDef};
    use quote::quote;
    use std::borrow::Cow;
    use std::collections::HashSet;

    #[test]
    fn test_opaque_gen() {
        let tokens = quote! {
            #[diplomat::bridge]
            #[diplomat::attr(auto, namespace = "mylib")]
            mod ffi {

                #[diplomat::opaque]
                struct OpaqueStruct;

                impl OpaqueStruct {
                    pub fn new() -> Box<OpaqueStruct> {
                        Box::new(OpaqueStruct{})
                    }

                    pub fn do_thing() -> bool {
                        return true;
                    }
                }
            }
        };
        let item = syn::parse2::<syn::File>(tokens).expect("failed to parse item ");

        let mut attr_validator = hir::BasicAttributeValidator::new("python");
        attr_validator.support = crate::nanobind::attr_support();

        let tcx = match hir::TypeContext::from_syn(&item, attr_validator) {
            Ok(context) => context,
            Err(e) => {
                for (_cx, err) in e {
                    eprintln!("Lowering error: {}", err);
                }
                panic!("Failed to create context")
            }
        };

        let (type_id, opaque_def) = match tcx
            .all_types()
            .next()
            .expect("Failed to generate first opaque def")
        {
            (type_id, TypeDef::Opaque(opaque_def)) => (type_id, opaque_def),
            _ => panic!("Failed to find opaque type from AST"),
        };

        let formatter = crate::nanobind::PyFormatter::new(&tcx);
        let errors = crate::ErrorStore::default();
        let mut binding = crate::nanobind::Binding::new();
        binding.module_name = std::borrow::Cow::Borrowed("pymod");

        let decl_header_path = formatter.fmt_decl_header_path(type_id);
        let impl_file_path = formatter.fmt_impl_file_path(type_id);

        let mut submodules = HashSet::<Cow<str>>::new();
        let mut context = crate::nanobind::TyGenContext {
            formatter: &formatter,
            errors: &errors,
            c2: crate::c::TyGenContext {
                tcx: &tcx,
                formatter: &formatter.c,
                errors: &errors,
                is_for_cpp: false,
                id: type_id.into(),
                decl_header_path: &decl_header_path,
                impl_header_path: &impl_file_path,
            },
            binding: &mut binding,
            generating_struct_fields: false,
            submodules: &mut submodules,
        };

        context.gen_opaque_def(opaque_def, type_id);
        let generated = binding.to_string();
        insta::assert_snapshot!(generated)
    }

    #[test]
    fn test_enum_gen() {
        let tokens = quote! {
            #[diplomat::bridge]
            #[diplomat::attr(auto, namespace = "mylib")]
            mod ffi {

                #[diplomat::enum_convert(my_thingy::SpeedSetting)]
                pub enum SpeedSetting {
                    Fast, Medium, Slow
                }
            }
        };
        let item = syn::parse2::<syn::File>(tokens).expect("failed to parse item ");

        let mut attr_validator = hir::BasicAttributeValidator::new("python");
        attr_validator.support = crate::nanobind::attr_support();

        let tcx = match hir::TypeContext::from_syn(&item, attr_validator) {
            Ok(context) => context,
            Err(e) => {
                for (_cx, err) in e {
                    eprintln!("Lowering error: {}", err);
                }
                panic!("Failed to create context")
            }
        };

        let (type_id, enum_def) = match tcx
            .all_types()
            .next()
            .expect("Failed to generate first opaque def")
        {
            (type_id, TypeDef::Enum(enum_def)) => (type_id, enum_def),
            _ => panic!("Failed to find opaque type from AST"),
        };

        let formatter = crate::nanobind::PyFormatter::new(&tcx);
        let errors = crate::ErrorStore::default();
        let mut binding = crate::nanobind::Binding::new();
        binding.module_name = std::borrow::Cow::Borrowed("pymod");

        let decl_header_path = formatter.fmt_decl_header_path(type_id);
        let impl_file_path = formatter.fmt_impl_file_path(type_id);

        let mut submodules = HashSet::<Cow<str>>::new();
        let mut context = crate::nanobind::TyGenContext {
            formatter: &formatter,
            errors: &errors,
            c2: crate::c::TyGenContext {
                tcx: &tcx,
                formatter: &formatter.c,
                errors: &errors,
                is_for_cpp: false,
                id: type_id.into(),
                decl_header_path: &decl_header_path,
                impl_header_path: &impl_file_path,
            },
            binding: &mut binding,
            generating_struct_fields: false,
            submodules: &mut submodules,
        };

        context.gen_enum_def(enum_def, type_id);
        let generated = binding.to_string();
        insta::assert_snapshot!(generated)
    }

    #[test]
    fn test_struct_gen() {
        let tokens = quote! {
            #[diplomat::bridge]
            #[diplomat::attr(auto, namespace = "mylib")]
            mod ffi {
                pub struct Thingy {
                    pub a: bool,
                    pub b: u8,
                    pub c: f64,
                }
            }
        };
        let item = syn::parse2::<syn::File>(tokens).expect("failed to parse item ");

        let mut attr_validator = hir::BasicAttributeValidator::new("python");
        attr_validator.support = crate::nanobind::attr_support();

        let tcx = match hir::TypeContext::from_syn(&item, attr_validator) {
            Ok(context) => context,
            Err(e) => {
                for (_cx, err) in e {
                    eprintln!("Lowering error: {}", err);
                }
                panic!("Failed to create context")
            }
        };

        let (type_id, struct_def) = match tcx
            .all_types()
            .next()
            .expect("Failed to generate first opaque def")
        {
            (type_id, TypeDef::Struct(struct_def)) => (type_id, struct_def),
            _ => panic!("Failed to find opaque type from AST"),
        };

        let formatter = crate::nanobind::PyFormatter::new(&tcx);
        let errors = crate::ErrorStore::default();
        let mut binding = crate::nanobind::Binding::new();
        binding.module_name = std::borrow::Cow::Borrowed("pymod");

        let decl_header_path = formatter.fmt_decl_header_path(type_id);
        let impl_file_path = formatter.fmt_impl_file_path(type_id);

        let mut submodules = HashSet::<Cow<str>>::new();
        let mut context = crate::nanobind::TyGenContext {
            formatter: &formatter,
            errors: &errors,
            c2: crate::c::TyGenContext {
                tcx: &tcx,
                formatter: &formatter.c,
                errors: &errors,
                is_for_cpp: false,
                id: type_id.into(),
                decl_header_path: &decl_header_path,
                impl_header_path: &impl_file_path,
            },
            binding: &mut binding,
            generating_struct_fields: false,
            submodules: &mut submodules,
        };

        context.gen_struct_def(struct_def, type_id);
        let generated = binding.to_string();
        insta::assert_snapshot!(generated)
    }
}
