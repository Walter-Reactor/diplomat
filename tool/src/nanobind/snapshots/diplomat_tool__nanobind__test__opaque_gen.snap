---
source: tool/src/nanobind/mod.rs
expression: generated
---
#include <nanobind/nanobind.h>
#include <nanobind/operators.h>
#include <nanobind/stl/unique_ptr.h>
#include <nanobind/stl/string_view.h>
#include <nanobind/stl/optional.h>
#include <nanobind/stl/function.h>
#include <../src/nb_internals.h>  // Required for shimming
#include "mylib/OpaqueStruct.hpp"

namespace nb = nanobind;
using namespace nb::literals;

// Nanobind does not usually support custom deleters, so we're shimming some of the machinery to add that ability.
// On module init, the dummy type will have the normal nanobind inst_dealloc function in the tp_dealloc slot, so we
// pull it out, store it here, and then call it in the tp_dealloc function we are shimming in to all our types.
// Our custom tp_dealloc function will call the tp_free function instead of `delete`, allowing us effectively to override
// the delete operator.
// See https://nanobind.readthedocs.io/en/latest/lowlevel.html#customizing-type-creation and
// https://github.com/wjakob/nanobind/discussions/932
static void (*nb_tp_dealloc)(void *) = nullptr;

void diplomat_tp_dealloc(PyObject *self)
{
    using namespace nb::detail;
    PyTypeObject *tp = Py_TYPE(self);
    const type_data *t = nb_type_data(tp);

    nb_inst *inst = (nb_inst *)self;
    void *p = inst_ptr(inst);
    if (inst->destruct)
    {
        inst->destruct = false;
        check(t->flags & (uint32_t)type_flags::is_destructible,
              "nanobind::detail::inst_dealloc(\"%s\"): attempted to call "
              "the destructor of a non-destructible type!",
              t->name);
        if (t->flags & (uint32_t)type_flags::has_destruct)
            t->destruct(p);
    }
    if (inst->cpp_delete)
    {
        inst->cpp_delete = false;
        auto tp_free = (freefunc)(PyType_GetSlot(tp, Py_tp_free));
        (*tp_free)(p);
    }
    (*nb_tp_dealloc)(self);
}

struct _Dummy {};

// Nanobind does not ship with support for casting char32_t, which seems to be an oversight.
// Remove this block when upstream support is added
namespace nanobind::detail
{
    template <>
    struct type_caster<char32_t>
    {
        using Value = char32_t;
        Value value;
        Py_ssize_t size;
        static constexpr auto Name = const_name("str");
        template <typename T>
        using Cast = char32_t;

        bool from_python(handle src, uint8_t, cleanup_list *) noexcept
        {
            value = PyUnicode_ReadChar(src.ptr(), 0);
            if (!value)
            {
                PyErr_Clear();
                return false;
            }
            size = PyUnicode_GetLength(src.ptr());
            return true;
        }

        static handle from_cpp(const char32_t *value, rv_policy,
                               cleanup_list *) noexcept
        {
            if (value == nullptr)
            {
                PyObject *result = Py_None;
                Py_INCREF(result);
                return result;
            }
            size_t len = 0;
            const char32_t *str = value;
            while (*str != U'\0')
            {
                len++;
                str++;
            }
            return PyUnicode_DecodeUTF32(reinterpret_cast<const char *>(value), len * 4, nullptr, nullptr);
        }

        static handle from_cpp(char32_t value, rv_policy, cleanup_list *) noexcept
        {
            return PyUnicode_DecodeUTF32(reinterpret_cast<const char *>(&value), 4, nullptr, nullptr);
        }

        template <typename T_>
        NB_INLINE bool can_cast() const noexcept
        {
            return (value && size == 1);
        }

        explicit operator char32_t()
        {
            if (can_cast<char32_t>())
                return value;
            else
                throw next_overload();
        }
    };

    template <typename T, typename E>
	struct type_caster<diplomat::result<T, E>>
	{
		using Value = diplomat::result<T, E>;
		Value value;
		Py_ssize_t size;
		using Caster = make_caster<T>;
		static constexpr auto Name = Caster::Name;


		static handle from_cpp(diplomat::result<T, E> value, rv_policy p, cleanup_list *cl) noexcept
		{
			if (value.is_ok()) {
				return Caster::from_cpp(forward_like_<T>(std::move(value).ok().value()), p, cl);
			}

			auto errorPyV = nb::cast(std::move(std::move(value).err().value()));
			if (errorPyV.is_valid())
			{
				PyErr_SetString(PyExc_Exception, nb::str(errorPyV).c_str());
			}
			else
			{
				char error_msg[512];
				snprintf(error_msg, sizeof(error_msg), "Cannot convert unknown type %s to string for python error.", typeid(E).name());
				PyErr_SetString(PyExc_Exception, error_msg);
			}

            return nullptr;
		}

		NB_INLINE bool can_cast() const noexcept { return Caster::template can_cast<T>(); }
	};
}


NB_MODULE(pymod, pymod_mod)
{
	{
		nb::class_<_Dummy> dummy(somelib_mod, "__dummy__");
		nb_tp_dealloc = (void (*)(void *))nb::type_get_slot(dummy, Py_tp_dealloc);
	}

    nb::class_<std::monostate>(somelib_mod, "monostate")
		.def("__repr__", [](const std::monostate &)
			 { return "()"; })
		.def("__str__", [](const std::monostate &)
			 { return "()"; });
    
    nb::module_ mylib_mod = pymod_mod.def_submodule("mylib");
    
    PyType_Slot mylib_OpaqueStruct_slots[] = {
        {Py_tp_free, (void *)mylib::OpaqueStruct::operator delete },
        {Py_tp_dealloc, (void *)diplomat_tp_dealloc},
        {0, nullptr}};
    
    nb::class_<mylib::OpaqueStruct>(mylib_mod, "OpaqueStruct", nb::type_slots(mylib_OpaqueStruct_slots))
    	.def_static("new_", &mylib::OpaqueStruct::new_)
    	.def_static("do_thing", &mylib::OpaqueStruct::do_thing);
}
